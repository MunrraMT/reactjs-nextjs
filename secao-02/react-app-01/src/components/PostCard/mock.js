const postCardPropsMock = {
  title: 'title 01',
  body: 'body 01',
  cover: 'img/img.webp'
};

export default postCardPropsMock;
