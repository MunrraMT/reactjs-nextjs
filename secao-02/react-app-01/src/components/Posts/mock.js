const postsPropsMock = [
  {
    title: 'title 01',
    body: 'body 01',
    cover: 'img/img1.webp',
    id: 1
  },
  {
    title: 'title 02',
    body: 'body 02',
    cover: 'img/img2.webp',
    id: 2
  }
];

export default postsPropsMock;
