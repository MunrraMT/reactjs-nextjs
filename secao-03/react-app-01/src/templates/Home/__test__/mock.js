const posts = [
  {
    userId: 1,
    id: 1,
    title: 'title 01',
    body: 'body 01',
  },
  {
    userId: 1,
    id: 2,
    title: 'title 02',
    body: 'body 02',
  },
  {
    userId: 1,
    id: 3,
    title: 'title 03',
    body: 'body 03',
  },
  {
    userId: 1,
    id: 4,
    title: 'title 04',
    body: 'body 04',
  },
  {
    userId: 1,
    id: 5,
    title: 'title 05',
    body: 'body 05',
  },
  {
    userId: 1,
    id: 6,
    title: 'title 06',
    body: 'body 06',
  },
];

const photos = [
  {
    albumId: 1,
    id: 1,
    title: 'title img 01',
    url: 'https://via.placeholder.com/600/92c952',
    thumbnailUrl: 'https://via.placeholder.com/150/92c952',
  },
  {
    albumId: 1,
    id: 2,
    title: 'title img 02',
    url: 'https://via.placeholder.com/600/771796',
    thumbnailUrl: 'https://via.placeholder.com/150/771796',
  },
  {
    albumId: 1,
    id: 3,
    title: 'title img 03',
    url: 'https://via.placeholder.com/600/24f355',
    thumbnailUrl: 'https://via.placeholder.com/150/24f355',
  },
  {
    albumId: 1,
    id: 4,
    title: 'title img 04',
    url: 'https://via.placeholder.com/600/92c952',
    thumbnailUrl: 'https://via.placeholder.com/150/92c952',
  },
  {
    albumId: 1,
    id: 5,
    title: 'title img 05',
    url: 'https://via.placeholder.com/600/771796',
    thumbnailUrl: 'https://via.placeholder.com/150/771796',
  },
  {
    albumId: 1,
    id: 6,
    title: 'title img 06',
    url: 'https://via.placeholder.com/600/24f355',
    thumbnailUrl: 'https://via.placeholder.com/150/24f355',
  },
];

export { posts, photos };
