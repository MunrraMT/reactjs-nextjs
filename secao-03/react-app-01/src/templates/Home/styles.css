.container {
  min-height: calc(100vh - 10rem);
  background-color: #eee;
  padding: 1rem;
}
