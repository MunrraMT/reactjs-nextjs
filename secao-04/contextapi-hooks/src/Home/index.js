import { useState } from 'react';
import Button from '../components/Button';
import Header from '../components/Header';
import { useCounterContext } from '../contexts/CounterContext';

const Home = () => {
  const [state, actions] = useCounterContext();

  const [value, setValue] = useState('');

  return (
    <main>
      <Header />

      <Button onBtnClick={actions.increase}>Increase</Button>
      <Button onBtnClick={actions.decrease}>Decrease</Button>
      <Button onBtnClick={actions.reset}>Reset</Button>
    </main>
  );
};

export default Home;
