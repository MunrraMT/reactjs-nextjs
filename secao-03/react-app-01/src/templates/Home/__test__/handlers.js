// eslint-disable-next-line import/no-extraneous-dependencies
import { rest } from 'msw';
import { photos, posts } from './mock';

const handlers = [
  rest.get('https://jsonplaceholder.typicode.com/posts', (req, res, ctx) =>
    res(ctx.status(200), ctx.json(posts)),
  ),

  rest.get('https://jsonplaceholder.typicode.com/photos', (req, res, ctx) =>
    res(ctx.status(200), ctx.json(photos)),
  ),
];

export default handlers;
