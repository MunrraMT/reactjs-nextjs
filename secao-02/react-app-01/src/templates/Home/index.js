import './styles.css';

import { useEffect, useState } from 'react';

import loadPosts from '../../utils/load-posts';
import Posts from '../../components/Posts';
import Button from '../../components/Button';
import TextInput from '../../components/TextInput';

const Home = () => {
  const [posts, setPosts] = useState([]);
  const [allPosts, setaAllPosts] = useState([]);
  const [page, setPage] = useState(0);
  const [postsPerPage] = useState(3);
  const [searchValue, setSearchValue] = useState('');

  useEffect(() => {
    const handleLoadPosts = async () => {
      const postsAndPhotos = await loadPosts();

      setPosts(postsAndPhotos.slice(page, postsPerPage));
      setaAllPosts(postsAndPhotos);
    };

    handleLoadPosts();
  }, []);

  const loadMorePosts = () => {
    const nextPage = page + postsPerPage;
    const nextPosts = allPosts.slice(nextPage, nextPage + postsPerPage);

    setPosts([...posts, ...nextPosts]);
    setPage(nextPage);
  };

  const handleInput = (e) => {
    const { value } = e.target;
    setSearchValue(value);
  };

  const noMorePosts = posts.length === allPosts.length;

  const filteredPosts = allPosts.filter((post) => {
    return (
      post.title.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1 ||
      post.body.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1
    );
  });

  return (
    <>
      <header
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          padding: '1rem',
          height: '7rem'
        }}
      >
        <h1>Estudo de ReactJS - App Posts</h1>
        <TextInput searchValue={searchValue} handleInput={handleInput} />
      </header>

      <main className='container'>
        {searchValue ? (
          (filteredPosts.length === 0 && <p>Busca não encontrada</p>) ||
          (filteredPosts.length > 0 && <Posts posts={filteredPosts} />)
        ) : (
          <Posts posts={posts} />
        )}
      </main>

      {!searchValue && (
        <footer>
          <Button disabled={noMorePosts} handleClick={loadMorePosts}>
            Ver mais posts
          </Button>
        </footer>
      )}
    </>
  );
};

// class Home2 extends Component {
//   state = {
//     posts: [],
//     allPosts: [],
//     page: 0,
//     postsPerPage: 3,
//     searchValue: ''
//   };

//   loadPosts = async () => {
//     const { page, postsPerPage } = this.state;

//     const postsAndPhotos = await loadPosts();

//     this.setState({
//       posts: postsAndPhotos.slice(page, postsPerPage),
//       allPosts: postsAndPhotos
//     });
//   };

//   loadMorePosts = () => {
//     const { page, postsPerPage, allPosts, posts } = this.state;

//     const nextPage = page + postsPerPage;
//     const nextPosts = allPosts.slice(nextPage, nextPage + postsPerPage);

//     posts.push(...nextPosts);

//     this.setState({ posts, page: nextPage });
//   };

//   handleInput = (e) => {
//     const { value } = e.target;
//     this.setState({ searchValue: value });
//   };

//   async componentDidMount() {
//     await this.loadPosts();
//   }

//   componentDidUpdate() {
//     console.log(this.props['test-props']);
//   }

//   render() {
//     const { posts, allPosts, searchValue } = this.state;

//     const noMorePosts = posts.length === allPosts.length;

//     const filteredPosts = allPosts.filter((post) => {
//       return (
//         post.title.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1 ||
//         post.body.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1
//       );
//     });

//     return (
//       <>
//         <header
//           style={{
//             display: 'flex',
//             flexDirection: 'column',
//             alignItems: 'center',
//             justifyContent: 'center',
//             padding: '1rem',
//             height: '7rem'
//           }}
//         >
//           <h1>Estudo de ReactJS - App Posts</h1>
//           <TextInput searchValue={searchValue} handleInput={this.handleInput} />
//         </header>

//         <main className='container'>
//           {searchValue ? (
//             (filteredPosts.length === 0 && <p>Busca não encontrada</p>) ||
//             (filteredPosts.length > 0 && <Posts posts={filteredPosts} />)
//           ) : (
//             <Posts posts={posts} />
//           )}
//         </main>

//         {!searchValue && (
//           <footer>
//             <Button disabled={noMorePosts} handleClick={this.loadMorePosts}>
//               Ver mais posts
//             </Button>
//           </footer>
//         )}
//       </>
//     );
//   }
// }

// function App() {
//   return (
//     <div className="App">
//       <header className="App-header">
//         <img src={logo} className="App-logo" alt="logo" />
//         <p>
//           Edit <code>src/App.js</code> and save to reload.
//         </p>
//         <a
//           className="App-link"
//           href="https://reactjs.org"
//           target="_blank"
//           rel="noopener noreferrer"
//         >
//           Learn React
//         </a>
//       </header>
//     </div>
//   );
// }

export default Home;
