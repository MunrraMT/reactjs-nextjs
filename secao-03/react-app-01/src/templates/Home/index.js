import './styles.css';

import { useEffect, useState } from 'react';

import loadPosts from '../../utils/load-posts';
import Posts from '../../components/Posts';
import Button from '../../components/Button';
import TextInput from '../../components/TextInput';

const Home = () => {
  const [posts, setPosts] = useState([]);
  const [allPosts, setaAllPosts] = useState([]);
  const [page, setPage] = useState(0);
  const [postsPerPage] = useState(3);
  const [searchValue, setSearchValue] = useState('');

  useEffect(() => {
    let isCancelled = false;

    loadPosts().then((postsAndPhotos) => {
      if (!isCancelled) {
        setPosts(postsAndPhotos.slice(page, postsPerPage));
        setaAllPosts(postsAndPhotos);
      }
    });

    return () => {
      isCancelled = true;
    };
  }, []);

  const loadMorePosts = () => {
    const nextPage = page + postsPerPage;
    const nextPosts = allPosts.slice(nextPage, nextPage + postsPerPage);

    setPosts((prevPosts) => [...prevPosts, ...nextPosts]);
    setPage(nextPage);
  };

  const handleInput = (e) => {
    const { value } = e.target;
    setSearchValue(value);
  };

  const noMorePosts = posts.length === allPosts.length;

  const filteredPosts = allPosts.filter(
    (post) =>
      post.title.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1,
  );

  return (
    <>
      <header
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          padding: '1rem',
          height: '7rem',
        }}
      >
        <h1>Estudo de ReactJS - App Posts</h1>
        <TextInput searchValue={searchValue} handleInput={handleInput} />
      </header>

      <main className='container'>
        {searchValue ? (
          (filteredPosts.length === 0 && <p>Busca n√£o encontrada</p>) ||
          (filteredPosts.length > 0 && <Posts posts={filteredPosts} />)
        ) : (
          <Posts posts={posts} />
        )}
      </main>

      {!searchValue && (
        <footer>
          <Button
            disabled={noMorePosts}
            handleClick={loadMorePosts}
            text='Ver mais posts'
          />
        </footer>
      )}
    </>
  );
};

export default Home;
