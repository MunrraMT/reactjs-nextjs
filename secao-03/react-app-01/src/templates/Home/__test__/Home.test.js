import {
  render,
  waitForElementToBeRemoved,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { setupServer } from 'msw/node';

import Home from '..';
import handlers from './handlers';
import { photos, posts } from './mock';

describe('Deve testar o componente Home', () => {
  const server = setupServer(...handlers);

  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  test('Deve renderizar sem erros', () => {
    render(<Home />);
  });

  test('Deve renderizar um título - "Estudo de ReactJS - App Posts"', () => {
    const { getByRole } = render(<Home />);

    expect(getByRole('heading')).toHaveTextContent(
      'Estudo de ReactJS - App Posts',
    );
  });

  test('Deve renderizar um texto dizendo "Loading..." quando não ouver posts', () => {
    const { getByText } = render(<Home />);

    expect(getByText('Loading...')).toBeInTheDocument();
  });

  test('Deve renderizar um texto dizendo "Loading..." até que os posts sejam recebidos pela API Externa', async () => {
    const { getByText } = render(<Home />);

    const message = getByText('Loading...');

    expect(message).toBeInTheDocument();

    await waitForElementToBeRemoved(message);
  });

  test('Deve renderizar os <PostCard /> pelo componente <Posts />, quando receber os dados da API Externa', async () => {
    const { getAllByRole, getByText, getByAltText } = render(<Home />);

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(3);

      expect(getByText(posts[0].title)).toBeInTheDocument();
      expect(getByText(posts[0].body)).toBeInTheDocument();
      expect(getByAltText(posts[0].title)).toHaveAttribute(
        'src',
        photos[0].url,
      );

      expect(getByText(posts[1].title)).toBeInTheDocument();
      expect(getByText(posts[1].body)).toBeInTheDocument();
      expect(getByAltText(posts[1].title)).toHaveAttribute(
        'src',
        photos[1].url,
      );

      expect(getByText(posts[2].title)).toBeInTheDocument();
      expect(getByText(posts[2].body)).toBeInTheDocument();
      expect(getByAltText(posts[2].title)).toHaveAttribute(
        'src',
        photos[2].url,
      );
    });
  });

  test('Deve renderizar o componente <TextIpunt />', () => {
    const { getByPlaceholderText } = render(<Home />);

    expect(getByPlaceholderText('Busque aqui')).toBeInTheDocument();
  });

  test('Deve pesquisar por posts através do componente <TextInput />', async () => {
    const { getAllByRole, getByPlaceholderText, getByText, getByAltText } =
      render(<Home />);

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(3);
    });

    userEvent.type(getByPlaceholderText('Busque aqui'), posts[1].title);

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(1);

      expect(getByText(posts[1].title)).toBeInTheDocument();
      expect(getByText(posts[1].body)).toBeInTheDocument();
      expect(getByAltText(posts[1].title)).toHaveAttribute(
        'src',
        photos[1].url,
      );
    });
  });

  test('Deve renderizar o texto "Busca não encontrada" quando não encontrar o texto pesquisado pelo componente <TextInput />', async () => {
    const { getAllByRole, queryAllByRole, getByPlaceholderText, getByText } =
      render(<Home />);

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(3);
    });

    userEvent.type(getByPlaceholderText('Busque aqui'), 'Não vai encontrar');

    await waitFor(() => {
      expect(queryAllByRole('article')).toHaveLength(0);
      expect(getByText('Busca não encontrada')).toBeInTheDocument();
    });
  });

  test('Deve renderizar o componente <Button /> com o texto "Ver mais posts"', () => {
    const { getByRole } = render(<Home />);

    expect(getByRole('button')).toHaveTextContent('Ver mais posts');
  });

  test('Deve mostrar mais posts quando o componente <Button /> for clicado', async () => {
    const { getAllByRole, getByRole } = render(<Home />);

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(3);
    });

    userEvent.click(getByRole('button'));

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(6);
    });
  });

  test('Deve fazer um Snapshot', async () => {
    const { getAllByRole, container } = render(<Home />);

    await waitFor(() => {
      expect(getAllByRole('article')).toHaveLength(3);
      expect(container).toMatchSnapshot();
    });
  });
});
