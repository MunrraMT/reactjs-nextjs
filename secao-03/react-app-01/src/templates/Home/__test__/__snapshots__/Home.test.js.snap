// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deve testar o componente Home Deve fazer um Snapshot 1`] = `
<div>
  <header
    style="display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 1rem; height: 7rem;"
  >
    <h1>
      Estudo de ReactJS - App Posts
    </h1>
    <input
      class="text-input"
      placeholder="Busque aqui"
      type="search"
      value=""
    />
  </header>
  <main
    class="container"
  >
    <section
      class="posts"
    >
      <article
        class="post"
      >
        <header>
          <img
            alt="title 01"
            class="post__img"
            src="https://via.placeholder.com/600/92c952"
          />
        </header>
        <section
          class="post__content"
        >
          <h2
            style="margin-bottom: 1rem;"
          >
            title 01
          </h2>
          <p>
            body 01
          </p>
        </section>
      </article>
      <article
        class="post"
      >
        <header>
          <img
            alt="title 02"
            class="post__img"
            src="https://via.placeholder.com/600/771796"
          />
        </header>
        <section
          class="post__content"
        >
          <h2
            style="margin-bottom: 1rem;"
          >
            title 02
          </h2>
          <p>
            body 02
          </p>
        </section>
      </article>
      <article
        class="post"
      >
        <header>
          <img
            alt="title 03"
            class="post__img"
            src="https://via.placeholder.com/600/24f355"
          />
        </header>
        <section
          class="post__content"
        >
          <h2
            style="margin-bottom: 1rem;"
          >
            title 03
          </h2>
          <p>
            body 03
          </p>
        </section>
      </article>
    </section>
  </main>
  <footer>
    <button
      class="btn-more-posts"
      type="button"
    >
      Ver mais posts
    </button>
  </footer>
</div>
`;
